File: src/tests/sz.rs
use std::fs::File;
use std::io::{BufRead, BufReader};
use tabled::{settings::Panel, settings::Style, Table};
use tokenizer_py::{tokenize, Token};
use walkdir::WalkDir;

#[test]
fn sz() -> std::io::Result<()> {
    let mut table: Vec<(String, i32, f32)> = vec![];

    for entry in WalkDir::new("src/teenygrad") {
        let filepath = entry.unwrap().path().to_string_lossy().to_string();
        if filepath.ends_with(".rs") {
            let tokens = BufReader::new(File::open(&filepath).unwrap())
                .lines()
                .filter_map(|line| line.ok())
                .into_iter()
                .map(|item| {
                    tokenize(item.to_string())
                        .unwrap()
                        .into_iter()
                        .filter(|token| {
                            matches!(
                                token,
                                Token::Name(_)
                                    | Token::OP(_)
                                    | Token::Number(_)
                                    | Token::String(_)
                                    | Token::Comment(_)
                            )
                        })
                        .take_while(|token| !matches!(token, Token::OP(s) if s == "//"))
                        .collect::<Vec<_>>()
                })
                .filter(|vec| !vec.is_empty())
                .collect::<Vec<Vec<_>>>();
            table.push((
                filepath.chars().skip(4).collect::<String>(),
                tokens.len() as i32,
                (tokens.iter().map(|vec| vec.len()).sum::<usize>() as f32) / (tokens.len() as f32),
            ));
        }
    }

    println!(
        "\n{}",
        Table::new(&table).with(Style::psql()).with(Panel::header(
            "NAME                                 | LINES        | TOKENS/LINE"
        ))
    ); // I really wish it was nicer to make this table

    let total_lines = table.iter().map(|row| row.1).sum::<i32>();
    println!("\nTotal line count: {:?}", total_lines);
    assert!(total_lines < 1500, "COME ON BRUV, ITS SUPPOSE TO BE TINY");
    return Ok(());
}

File: src/tests/sz.py
"""
Did my own implementation sz.rs, but the dude that had tokenize_py cargo package deleted it. 

I am quite upset about it since I was quite proud of that code. 

Added a line at 29 to break the tokenization process if it detects a single line comment, therefore we do not count rust comments in our tokens.

- Alex
"""

#!/usr/bin/env python3
import os
import token
import tokenize
import itertools
from tabulate import tabulate

TOKEN_WHITELIST = [token.OP, token.NAME, token.NUMBER, token.STRING, token.COMMENT]

if __name__ == "__main__":
  headers = ["Name", "Lines", "Tokens/Line"]
  table = []
  for path, subdirs, files in os.walk("src/teenygrad"):
    for name in files:
      if not name.endswith(".rs"): continue
      filepath = os.path.join(path, name)
      with tokenize.open(filepath) as file_:
        tokens = []
        for t in tokenize.generate_tokens(file_.readline):
            if t.string == "//": break
            if t.type in TOKEN_WHITELIST: tokens.append(t)
        token_count, line_count = len(tokens), len(set([t.start[0] for t in tokens]))
        table.append([filepath, line_count, 0]) # token_count/line_count])

  print(tabulate([headers] + sorted(table, key=lambda x: -x[1]), headers="firstrow", floatfmt=".1f")+"\n")

  for dir_name, group in itertools.groupby(sorted([(x[0].rsplit("/", 1)[0], x[1]) for x in table]), key=lambda x:x[0]):
    print(f"{dir_name:30s} : {sum([x[1] for x in group]):6d}")

  total_line_count = sum([x[1] for x in table])
  print(f"\ntotal line count: {total_line_count}")
  assert total_line_count < 5000, "TEENYGRAD-RS IS SUPPOSE TO BE TEENY TINY"
